### ISSUE TITLE
Napari Labels filesize too large

### ISSUE THREAD

**Issue by pranjaldhole** [napari/napari#2123]
[test_scenario.zip](https://github.com/napari/napari/files/5829137/test_scenario.zip)

## Feature request

- The label files generated by Napari has file sizes too large (larger than original dataset).
- For training the large volumetric data for image segmentation and classification, this is a problem since every time labeling is performed, the regenerated label files needs to be copied over to the cluster.
- Biomedical volumetric image data tend to be very large and this it is important that the napari label files have some form of compression implemented so that the label files are _portable_.

## To Reproduce

1. Take a few tiff files as your image data set (example data is attached with this issue).
2. Sparsely label some of the images to generate label classes.
3. Save all layers which produces output .tif files with input data and corresponding labels.

An example input dataset and output generated by Napari is attached with this issue to check the reproduce the aforementioned behaviour.

## Expected behavior

- *Option A*: A new function that can compress all the label tif files such that only marked images and regions are recorded. This file has a _portable_ size which can be copied to another machine and unpacked so that label tif files can be regenerated on another machine.
- *Option B*: Generate a napari label container file instead of label tif files (similar to how its done in ilastik software) that has relative paths written to input dataset and labels stored for images containing any labels. Each plugin will have to be able to read this file and unpack the labels so that training/testing can be performed.

## Additional context

[napari_write_labels](https://napari.org/docs/dev/plugins/hook_specifications.html#napari.plugins.hook_specifications.napari_write_labels) is the relevant hook specification related to this issue.

**Comment by sofroniewn** [comment]
> The label files generated by Napari has file sizes too large (larger than original dataset).

Right now it looks like we only do a minimal amount of dtype checking, see here:
https://github.com/napari/napari/blob/da4e309e5734931a7ecfc821ae84a33462d80a45/napari/plugins/_builtins.py#L115-L116 There might be an opportunity to do more conversions to things like `uint8`, `uint16` etc that will help a little. Curious your thoughts there @jni?

Stepping back, I know that @joshmoore and the OME/ Bioformats team has also been thinking about an efficient way to represent labels on disk, possibly using ome-zarr. My preference would be to team up with them to make that available in a standard io plugin which would then be available to all segmentation etc. plugins. Do you have any update on progress here @joshmoore?

**Comment by joshmoore** [comment]
The current version ([0.1](https://ngff.openmicroscopy.org/0.1/#label-md)) supports labeled images. There are methods in ome-zarr-py to write them but there's not yet a plugin. https://github.com/tlambert03/napari-omero/pull/17 adds support for writing them to _OMERO_ which can then download them to an OME-Zarr, so I'd imagine all the pieces are in place. It would be good to hear what you consider a "standard io plugin" though. Restrictions on imports? Installed by default? etc. etc.

**Comment by jni** [comment]
@sofroniewn 

>  There might be an opportunity to do more conversions to things like uint8, uint16 etc that will help a little.

I don't think the dtype is relevant at all, actually. The label files *should* be small, because labels compress like gangbusters, sparse labels even more so. So my guess is that our default writer plugin does *no* compression, and that fixing this will be enough to close this issue.

```python
In [1]: import numpy as np
In [2]: image = np.zeros((512, 512, 512), dtype=np.uint32)
In [3]: image.nbytes / 1e9
Out[3]: 0.536870912
In [4]: image[30:130, 30:130, 30:130] = 1
In [5]: image[250:300, 250:300, 250:300] = 2
In [6]: import tifffile
In [7]: tifffile.imsave('test.tif', image, compress=1)
In [8]: !ls -l test.tif
-rw-r--r--@ 1 jni  staff  2729120 19 Jan 22:43 test.tif
```

500MB -> 3MB compression.

**Comment by sofroniewn** [comment]
> The current version (0.1) supports labeled images ... It would be good to hear what you consider a "standard io plugin" though. Restrictions on imports? Installed by default? etc. etc.

Very nice, I didn't realize that! Very nice. I meant by "standard" one that is useable by others. I think wiring up the hook implementations is probably a good first step for napari

> I don't think the dtype is relevant at all, actually. The label files should be small, because labels compress like gangbusters, sparse labels even more so. So my guess is that our default writer plugin does no compression, and that fixing this will be enough to close this issue.

Ah yes, that's great too! I actually don't know too much about the compress options for tifffile, but that seems like a great place to start for @pranjaldhole

**Comment by pranjaldhole** [comment]
> Ah yes, that's great too! I actually don't know too much about the compress options for tifffile, but that seems like a great place to start for @pranjaldhole

@sofroniewn and @jni, I agree. `tifffile` compression should already significantly reduce the label file size.

**Collecting labels together**: 
The additional thing to do here would be to write a napari container file that collects all the labelled and unlabelled files and write a flag `UNLABELLED` for the unlabelled images while storing the compressed label files.

**Connecting labels to image dataset**:
From there we can associate the entire image data set and the associated labels that would also address [issue12](https://github.com/chanzuckerberg/napari-plugin-alfa-cohort/issues/12) and [issue19](https://github.com/chanzuckerberg/napari-plugin-alfa-cohort/issues/19).

Have you additional thoughts @cmohl2013?

**Comment by cmohl2013** [comment]
@pranjaldhole

I would not go for *Option B* (as we do for the current YAPiC command line interface). Because if you move your image and label data to another machine you will likely change the paths which will lead to import errors. Safer and more user-friendly is a single file, containing both labels and pixels. It is generally a good idea to exclude images without any labels. But you should not exclude unlabeled image regions per se. Unlabelled image regions next to labelled regions may be needed for training, depending on the network you train (the input layer is normally larger than the output layer). 

I think it is worth to have a look at the ome-zarr, looks very interesting and you would not have to build the container file from scratch.

**Comment by DragaDoncila** [comment]
@pranjaldhole I'm looking into a more portable label format for your use case, and I was hoping you could give me a bit of extra detail on your workflow. 

Your napari output, for example, has both labels layers stacked to correspond to your image stack. Is this because you chose to open the input images as a stack? I'm also curious about the two labels classes showing up as two separate labels layers, rather than one layer - is this your desired outcome, or would you prefer all classes in one layer?

If you could detail your workflow starting from how you open up the images (dragging in all of them at once or one by one, opening as a stack, opening using a script...) and new labels layers, and moving on to how you paint the labels and save the layers to file, it would give me a much better understanding of your process and hopefully inform a more functional portable format.

In terms of ome-zarr, I think the biggest disadvantage is the requirement for images and labels to be 5D, with singleton dimensions where that dimension is irrelevant. How big of an impact do you think this might have on your workflow?

**Comment by pranjaldhole** [comment]
@DragaDoncila Thanks for looking into this. Let me write down an expected workflow in the following steps:

**On User Laptop**:
1. I want to open a stack of images using the `Open Folder` function. This way all the images are loaded in Napari.
2. I will label only some of the images that I will choose manually from this stack. The images are labelled sparsely, i.e., only very small portion of the image is labelled.
3. I want to save all the labels (truly sparsely, i.e., only co-ordinates of labelled points in the labelled images are stored perhaps in the form of numpy arrays in HDF5 like storage) into a label container file that contains sparse labels and relative path to images so that labels can be corresponded with the images.
4. I copy this _portable_ label file to GPU machine.

**On GPU machine**:
1. I place the label data and the input images in the same folder so that relative paths are correct.
2. I train a NN model that can read this label container file (i.e., label file will be converted into a stack of binary numpy arrays matching the dimensions of the input images).
3. A keras model is generated after training which is then copied over to user laptop.

**On User Laptop**:
1. Predict unlabelled images with the trained model. Identify regions where model performs poorly.
2. Label more data so that regions where NN model performs poorly can be learnt better by the model.
3. Save the updated label file.

**On GPU machine**:
Retrain the model with updated label file.

In summary, I want to perform an iterative process of sparse labelling and model training for pixel classification. Therefore, I want to be able to copy this label file as often between User laptop and GPU machine as possible. _Portability_ of file (small size with all labels and relative paths to input data) is the key desired outcome here. Typically, I imagine the user to be an IT non-expert and will use USB sticks to move this file from one machine to another.

Let me know if the workflow is unclear somewhere. I'll try my best to clarify further. Thanks a lot!

**Comment by DragaDoncila** [comment]
@pranjaldhole thank you for all the detail on the workflow and your patience on this issue. I have written a plugin which I hope will be able to provide much of the functionality you are after and you can find it [here](https://github.com/DragaDoncila/napari-compressed-labels-io).

To mitigate the issue of compression, we've added compression to the default call for saving layers, so the tiffs should be a more reasonable size. The new plugin also has the option of saving out a stacked labels layer to zarr format, which compresses your oak labels example from 35MB (original napari output) to 245kB.

To improve portability and usability, I've added a writer which, when _Save All Layers_ is selected in the viewer, saves each slice of the image and label stacks into individual zarrs. This maintains a clear correspondence between each image and its associated labels, and also allows you to open just one slice at a time, or an individual label or image slice, as well as the full stack (using the readers provided in the plugin). While this method does not provide the best compression possible, it is still reasonable, with 10MB total output for your given example, as opposed to 83MB for the original napari output. I've attached this example output here.

[napari_output.zip](https://github.com/napari/napari/files/6026396/napari_output.zip)

To make use of this plugin you can install it via pip. Then, you can open the images as a stack, add your labels layer and paint your labels, and then _Save All Layers_ providing a filename with a `.zarr` extension e.g. `napari_output.zarr`. 
Note that the plugin only supports 3D layers at the moment, and is limited to `Labels` and `Image` layers. After it has saved you can then:
- Drag the full output folder into napari to open the full stack of all layers
- Drag a numbered subdirectory into napari to open the layers of a given slice
- Drag a named subdirectory within a numbered subdirectory into napari to open an individual image/label of a given slice

The plugin is brand new so please don't hesitate to let me know if you run into any issues (which may well be the case) on the [plugin repo itself](https://github.com/DragaDoncila/napari-compressed-labels-io). 
I'm sure there will be plenty of ways to improve the functionality, and indeed, change the format altogether if it's not what you were expecting/imagining/wanting. However, I thought it was a good starting base for further discussion and iterative improvements.

In broader terms, we will discuss the possibility of ome-zarr itself supporting similar functionality and providing writers to napari, but this is likely to be a longer process, and we wanted to make sure there was at least something to start with.

The issue of sparseness in Labels has yet to be addressed, but will be my focus now over the next week.

cc @neuromusic

**Comment by pranjaldhole** [comment]
@DragaDoncila Thank you for developing the plugin. I was testing the example file you uploaded. It seems that it is no longer possible for me to add further labeling to  the data. I get an error when I tried to add more labels as follows:

> Item assignment with <class 'tuple'> not supported

Is that an easy error to fix?

**Comment by DragaDoncila** [comment]
Hi @pranjaldhole , this should be fixed now. You'll need to upgrade your plugin version using pip install -U napari-compressed-labels-io.

Let me know if this works for you!

Note that the error was a result of zarr not supporting fancy indexing, which we need for painting. This means when we read the layers back, we must instantiate them as numpy arrays. Could this be an issue for you? How large are the datasets you usually work with?

**Comment by kcpevey** [comment]
@DragaDoncila @pranjaldhole Can we close this as complete?

**Comment by jni** [comment]
I think so! Thanks for helping triage @kcpevey! 🙏
